Userland/Libraries/LibTest/Randomized/Generator.h

```diff
@@ -0,0 +1,321 @@
+//                     etc.
+//
+// Shrinks towards 0.
+inline u32 unsigned_int()
+{
+    u32 bits = frequency(
+        // weight, bits
+        Choice { 4, 4 },
+        Choice { 8, 8 },
+        Choice { 2, 16 },
+        Choice { 1, 32 },
+        Choice { 2, 0 });
+
+    // The special cases go last as they can be the most extreme (large) values.
+
+    if (bits == 0) {
+        // Special cases, eg. max integers for u8, u16, u32.
+        return one_of(
+            0U,
+            NumericLimits<u8>::max(),
+            NumericLimits<u16>::max(),
+            NumericLimits<u32>::max());
+    }
+
+    u32 max = (bits == 32)
+        ? NumericLimits<u32>::max()
+        : ((u64)1 << bits) - 1;
+    return unsigned_int(max);
+}
+
+// A generator returning `true` with the given `probability` (0..1).
+//
+// If probability <= 0, doesn't use any randomness and returns false.
+// If probability >= 1, doesn't use any randomness and returns true.
+//
+// In general case:
+// Gen::weighted_boolean(0.75)
+//   -> value false, RandomRun [0]
+//   -> value true,  RandomRun [1]
+//
+// Shrinks towards false.
+inline bool weighted_boolean(double probability)
+{
+    if (probability <= 0)
+        return false;
+    if (probability >= 1)
+        return true;
+
+    u32 random_int = Test::randomness_source().draw_value(1, [&]() {
+        double drawn_probability = get_random_probability();
+        return drawn_probability <= probability ? 1 : 0;
+    });
+    bool random_bool = random_int == 1;
+    return random_bool;
+}
+
+// A (fair) boolean generator.
+//
+// Gen::boolean()
+//   -> value false, RandomRun [0]
+//   -> value true,  RandomRun [1]
+//
+// Shrinks towards false.
+inline bool boolean()
+{
+    return weighted_boolean(0.5);
+}
+
+// A vector generator of a random length between the given limits.
+//
+// Gen::vector(2,3,[]() { return Gen::unsigned_int(5); })
+//   -> value [1,5],      RandomRun [1,1,1,5,0]
+//   -> value [1,5,0],    RandomRun [1,1,1,5,1,0,0]
+//   etc.
+//
+// In case `min == max`, the RandomRun footprint will be smaller, as there will
+// be no randomness involved in figuring out the length:
+//
+// Gen::vector(3,3,[]() { return Gen::unsigned_int(5); })
+//   -> value [1,3], RandomRun [1,3]
+//   -> value [5,2], RandomRun [5,2]
+//   etc.
+//
+// Shrinks towards shorter vectors, with simpler elements inside.
+template<typename Fn>
+inline Vector<InvokeResult<Fn>> vector(size_t min, size_t max, Fn item_gen)
+{
+    VERIFY(max >= min);
+
+    size_t size = 0;
+    Vector<InvokeResult<Fn>> acc;
+
+    // Special case: no randomness for the boolean
+    if (min == max) {
+        while (size < min) {
+            acc.append(item_gen());
+            ++size;
+        }
+        return acc;
+    }
+
+    // General case: before each item we "flip a coin" to decide whether to
+    // generate another one.
+    while (size < min) {
+        acc.append(item_gen());
+        ++size;
+    }
```

Comment:
Is there a particular reason behind not just generating the final length instead (i.e. not using a geometric distribution, although I don't quite understand the {,purpose of the} probability math that follows below)? If so, might be worth noting down in the comment.

Reply:
Yes, the "flip a coin then generate/stop" approach produces RandomRuns that play nicer with the shrink commands we are trying.

The "generate a length then generate that many items" approach is so common that we have some shrink commands cater to it specifically, but it still will eg. only delete one item at a time, while the "flip a coin" approach can sometimes delete many items at once.

This is described better in [the paper](https://drops.dagstuhl.de/opus/volltexte/2020/13170/pdf/LIPIcs-ECOOP-2020-13.pdf), section 3.3, starting at "More commonly".

This is also discussed in the [TestGenerator.cpp](https://github.com/SerenityOS/serenity/pull/21191/files#diff-20c05541d3cc0bd0f347fae3291bcc6ab0c27c1bec353d84b3a4ad52d0531591R53-R88) file. I'll expand the comment in here (in Generator.h) as well though to discuss why not use the obvious approach.